LocalStorage/LocalStorageHelpers/RealmStorageHelper.swift

```diff
@@ -0,0 +1,131 @@
+//
+//  RealmStorageHelper.swift
+//  WeatherDemoApp
+//
+//  Created by ahmed on 05/04/2024.
+//
+
+import RealmSwift
+import Combine
+
+final class RealmStorageHelper: LocalStorageProtocol {
+    private var realm = try! Realm()
+    
+    private var notificationToken: NotificationToken!
+    private var citiesObservervable: Results<CityRealmObject>!
+    
+    /// outputs
+    let cities = PassthroughSubject<Result<[LocalStorageCity], LocalStorageError>, Never>()
+    
+    init() {
+        startObservingCities()
+    }
+    
+    deinit {
+        notificationToken?.invalidate()
+    }
+    
+    func getCitiesData() -> [LocalStorageCity] {
+        
+        return realm
+            .objects(CityRealmObject.self)
+            .sorted(by: \.cityName)
+            .map {$0.getLocalStorageCity()}
+    }
+    
+    // MARK: Add
+    /// add a city to local storage
+    func addCityWeatherInfoToLocalStorage(weatherInfo: FormattedCityWeatherModel) throws {
+        
+        /// if city is present in DB, then append history entity
+        if let city = queryCityByName(weatherInfo.cityName).first {
+            do {
+                try? realm.write {
+                    city.weatherInfoList.append(weatherInfo)
+                }
+            } catch {
+                throw LocalStorageError.addError
+            }
+            
+            return
+        }
+        
+        /// new city? then add the new city to DB
+        let city = CityRealmObject(weatherInfoItem: weatherInfo)
+        do{
+            try? realm.write {
+                realm.add(city)
+            }
+        } catch {
+            throw LocalStorageError.addError
+        }
+    }
+    
+    // MARK: Delete
+    func deleteCityFromLocalStorage(_ cityName: String) throws {
+        
+        if let city = queryCityByName(cityName).first {
```

Comment:
You should throw an error if you're unable to get the city. A nice fallback is to add it if it doesn't exist.